extensions/guacamole-auth-cas/src/main/java/org/apache/guacamole/auth/cas/ticket/TicketValidationService.java

```diff
@@ -122,16 +134,42 @@
             }
             
             // Convert remaining attributes that have values to Strings
+            String groupAttribute = confService.getGroupAttribute();
+            // Use cas-member-attribute to retrieve and set group memberships
+            String groupDnFormat = confService.getGroupDnFormat();
+            String groupTemplate = "";
+            if (groupDnFormat != null) {
+                // if CAS is backended to LDAP, groups come in as RFC4514 DN
+                // syntax.  If cas-group-dn-format is set, this strips
+                // an entry such as "CN=Foo,OU=Bar,DC=example,DC=com" to
+                // "Foo"
+                groupTemplate = groupDnFormat.replace("%s","([A-Za-z0-9_\\(\\)\\-\\.\\s+]+)");
+                // the underlying parser aggregates all instances of the same
+                // attribute, so we need to be able to parse them out
+                groupTemplate=groupTemplate+",*\\s*";
+            } else {
+               groupTemplate = "([A-Za-z0-9_\\(\\)\\-\\.\\s+]+,*\\s*)";
+            }
+            Pattern pattern = Pattern.compile(groupTemplate);
```

Comment:
What if the DN format contains a character that has special meaning to `Pattern.compile()`?

Reply:
This is the only feedback I have not yet addressed in commits. I am not sure if this is a problem or, if it is, how best to address it. I will look into this later this week if I have time. Please feel free to suggest or advise on best approach.

Reply:
I'm not 100% certain if this will work, but, based on this page:

https://www.baeldung.com/java-regexp-escape-char

It looks like the `\Q` and `\E` special characters can be used to escape anything between them. So it might be possible to use those around the value of the `groupDnFormat` String to automatically escape anything within that ??

Reply:
Although best practices would prohibit special characters in the DN it does not guarantee that. I believe your suggested use of \Q and \E to escape any special characters in the stripped DN is a good idea. Can we simply use the Pattern.quote() method as follows:

groupTemplate = Pattern.quote(groupDnFormat.replace("%s","([A-Za-z0-9_\\(\\)\\-\\.\\s+]+)"));

Or am I mistaken about where to escape the stripped DN?

Reply:
@mike-jumper Thoughts, here? RegEx still causes my mind to twist in ways I didn't think were possible...

![image](https://user-images.githubusercontent.com/4706000/99840624-e5e4b280-2b3a-11eb-98fb-c6f7f3dabbcf.png)

Reply:
Welp:

> It looks like the `\Q` and `\E` special characters can be used to escape anything between them. So it might be possible to use those around the value of the `groupDnFormat` String to automatically escape anything within that ??

Nope - that would be unsafe. What if the value of `groupDnFormat` were `"\E.*\Q"`?

> Can we simply use the `Pattern.quote()` method as follows:
>
>     groupTemplate = Pattern.quote(groupDnFormat.replace("%s","([A-Za-z0-9_\(\)\-\.\s+]+)"));

Not quite - that would escape the regex used as the replacement, as well. I think you have the right idea. Perhaps:

```java
groupTemplate = Pattern.quote(groupDnFormat).replace("%s","([A-Za-z0-9_\(\)\-\.\s+]+)");
```

BUT:

* Definitely have some [Leaning Toothpick Syndrome](https://en.wikipedia.org/wiki/Leaning_toothpick_syndrome) going on with that regex. It's really tough to validate in the current context. Probably best to move that to a constant that can be separately documented, understood, and verified... that way this magic turns into something like: `Pattern.quote(groupDnFormat).replace("%s", REGEX_THAT_WE_KNOW_IS_CORRECT)`
* I'm not 100% sure the regex _is_ correct ... depending on what it's supposed to match against (which I'm also not 100% sure of). Perhaps any added documentation will clarify that and allow for easier review.

Reply:
Also ... what if the DN format needs to contain a literal "%s"?

Reply:
Also:

* What if the relevant part of the DN in an actual group contains characters that must be escaped for the sake of LDAP (like a comma)?
* What if the relevant groups are nested in different subtrees of a common base DN, or involve different attributes?

It seems to me that the code transforming a DN received from CAS into a generic group name would need to be LDAP syntax-aware, rather than employing pattern matching.

Reply:
I think LdapName is the way:
http://cr.openjdk.java.net/~iris/se/12/latestSpec/api/java.naming/javax/naming/ldap/LdapName.html#%3Cinit%3E(java.lang.String)